### ISSUE TITLE
"invalid state received in the request" causing infinite loop on Login

### ISSUE THREAD

**Issue by securityvoid** [AzureAD/passport-azure-ad#247]
On a new login I'm seeing the error "Invalid state received in the request". 

I stepped through the break points and took a look at the code in oidcstrategy:
```
var tuple = self._sessionContentHandler.findAndDeleteTupleByState(req, self._key, params.state);
    if (!tuple)
      return self.failWithLog('In collectInfoFromReq: invalid state received in the request');
```

If I'm reading the code properly, it appears like its looking to see if there is a session for the user already and if there is not a session for the user already it throws the error. Since this is on the initial login before any other login, there isn't a session, therefore its kicking out the error. 

When the error is hit, it causes the authentication to fail which I have forward back to the sign-in page. The sign-in page tries to login to AzureAD. Since AzureAD is already logged into, it then forwards back to the application... which hits this error again and it repeats.

I can reproduce this bug 100% in one environment and only occasionally hit it in another environment with the same code. I'm not really sure why this is the case.

I have discovered if I change the code to:
```
    /*if (!tuple)
      return self.failWithLog('In collectInfoFromReq: invalid state received in the request');
      //*/
    if(tuple) {
      params.nonce = tuple['nonce'];
      params.policy = tuple['policy'];
      params.resource = tuple['resource'];
    }
```

And also change the following line to account for when optionsToValidateNonce is null:
```
    // check nonce
    if (!jwtClaims.nonce || jwtClaims.nonce === '' || (optionsToValidate.nonce && jwtClaims.nonce !== optionsToValidate.nonce))
```

Everything appears to work properly. 

However, since I'm not really clear why this error is happening, I'm having a difficult time evaluating if what I put together is a "good" fix. 

Can someone help clarify why I might be seeing this behavior, and if my fix is a good one?

I'm going against a Version 1 endpoint created in portal.azure.com.

**Comment by lovemaths** [comment]
@securityvoid When we send an authentication request to AAD, we always generate a {state, nonce, ...} tuple and save it into session; when we get the response from AAD, we retrieve the tuple using the state value, and remove the tuple from session. The information in the tuple is used for all kinds of validations. Therefore if we cannot get a tuple, something is wrong.

I don't know what goes wrong from the description. Can you try the following things and give me some more information? Thank you.

(1) Can you step into self._sessionContentHandler.findAndDeleteTupleByState, stop at line 49, print out 'array' and 'stateToFind' variables and see if there is tuple in 'array' that matches 'stateToFind'?

(2) We only keep the most recent 10 tuples (that were generated within the past hour) that are not handled. In other words, we only honor the most recent 10 login requests which haven't got a response from AAD yet. You can try to change the 'NONCE_MAX_AMOUNT' value at line 68 in oidcstrategy.js to a larger number, and see if this problem still occurs.

**Comment by lovemaths** [comment]
@securityvoid Please let me know if this issue still occurs. If so I will re-open this issue. Thanks.

**Comment by davidkpiano** [comment]
@lovemaths I'm seeing a similar issue, where I will get a `401 Unauthorized` and the page will infinitely be loading. It's best reproducible when using Edge (InPrivate window), if that helps.

**Comment by davidkpiano** [comment]
This seems related: https://stackoverflow.com/questions/40185796/trouble-using-passport-azure-ad-oidcstrategy-with-passportjs and I noticed that I had:

```js
app.use(bodyParser.urlencoded({extended: false}); // instead of true
```

which may be the issue re. inconsistent states.

**Comment by lovemaths** [comment]
@davidkpiano Yes that might be possible. If body parser fails to parse, then we cannot get the `state` parameter from the body. Please give it a try and let me know the result. Thank you.

**Comment by davidkpiano** [comment]
The same issue occurs when setting `extended: true`.

**Comment by HofmT** [comment]
@lovemaths Hi,
I have the same error. After I login for the first time using AAD I get redirected to the Login page of my WebApp and in the logs it prints out: _authentication failed due to: In collectInfoFromReq: invalid state received in the request_. Now, when I go to the login route a second time it will redirect me to the Homepage of the WebApp without entering my credentials again. Any help would be appreciated.

**Comment by davidkpiano** [comment]
@HofmT The issue for me was that there were multiple instances of the authentication service working across different servers, which led to an inconsistent (or missing `state`) when comparing the state received from AAD on the round-trip.

The best solution so far was to use cookies instead:

```js
passport.use(
  new OIDCStrategy(
    {
      // ...
      useCookieInsteadOfSession: true,
      cookieEncryptionKeys: [
        { key: COOKIE_ENCRYPTION_KEY, iv: COOKIE_ENCRYPTION_IV }
      ]
    },
    // ...
  )
);
```

**Comment by HofmT** [comment]
@davidkpiano Thanks for the swift reply! I managed to find my issue in the meantime as well. I'm currently testing the application on my localhost and I have registered the app in AAD under 127.0.0.1. When I access my page under _localhost_ I will end up receiving the state error. However, when accessing the page via 127.0.0.1 it will redirect me correctly. Such an akward issue, happy that I finally figured it out.
